### PR TITLE
PUPIL-388: Create OakPrimaryInvertedButton

### PR DESCRIPTION
**Pull request by weronika-szalas**
Create `OakPrimaryInvertedButton`:

Screenshot 2024-01-31 at 12 13 21

### In addition:
- rename `InternalRectButton` to `InternalShadowRectButton`
- rename `InternalRoundButton` to `InternalShadowRoundButton`
- add the `hoverShadow` prop to `InternalShadowRectButton` with  default value and allow it to be null if no hoverShadow wanted (needed for `OakPrimaryInvertedButton`)
- fix styles from styled components being overwritten by `"button-wrapper"` class in `InternalShadowRectButton` and `InternalShadowRoundButton` 

Storybook 👉🏼https://deploy-preview-90--lively-meringue-8ebd43.netlify.app/?path=/docs/components-ui-oakprimaryinvertedbutton--docs

Figma 👉🏼 https://www.figma.com/file/YcWQMMhHPVVmc47cHHEEAl/Oak-Design-Kit?type=design&node-id=3029-11819&mode=design

### CODE REVIEW

File: src/components/ui/OakPrimaryInvertedButton/OakPrimaryInvertedButton.stories.tsx [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+ *
+ * The following callbacks are available for tracking focus events:
+ *
+ * ### onClick
+ * `onClick?: (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;`
+ *
+ * ### onHovered
+ *  `onHovered?: (event: React.MouseEvent<HTMLButtonElement, MouseEvent>, duration: number) => void;`<br>
+ *  called after a mouseEnter and mouseLeave event has happened
+ */
```
Comment (carlmw) [inline src/components/ui/OakPrimaryInvertedButton/OakPrimaryInvertedButton.stories.tsx:23]: If this doc block is moved above the component itself here: https://github.com/oaknational/oak-components/blob/ce6f7e32f485dfc147c61651cbae7d7513070e00/src/components/ui/OakPrimaryInvertedButton/OakPrimaryInvertedButton.tsx#L22

then will show up in both Storybook and as a typescript hint in VSCode when using the component. Really helpful when consuming it in OWA.

We have these blocks in the stories in some places but i've been making this change with new components and want to do the same with older components.

File: src/components/ui/OakPrimaryInvertedButton/OakPrimaryInvertedButton.tsx [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  element,
+  ...rest
+}: OakPrimaryInvertedButtonProps & PolymorphicPropsWithoutRef<C>) => {
+  return (
+    <InternalShadowRectButton
+      element={element ?? "button"}
+      defaultBorderColor="bg-btn-secondary"
+      defaultBackground="bg-btn-secondary"
+      defaultTextColor="text-primary"
+      hoverBackground="bg-btn-secondary"
```
Comment (kimon-satan) [inline src/components/ui/OakPrimaryInvertedButton/OakPrimaryInvertedButton.tsx:32]: I don't understand how this isn't causing the background to change colour on hover ?

File: src/components/ui/OakPrimaryInvertedButton/OakPrimaryInvertedButton.tsx [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+      hoverBorderColor="bg-btn-secondary"
+      hoverTextColor="text-primary"
+      disabledBackground="bg-btn-secondary"
```
Comment (kimon-satan) [inline src/components/ui/OakPrimaryInvertedButton/OakPrimaryInvertedButton.tsx:35]: Same here - why doesn't the background render ?

File: src/components/ui/InternalShadowRectButton/InternalShadowRectButton.tsx [hunk 4]
```diff
@@ -80,16 +81,23 @@ const StyledInternalButton = styled(InternalButton)<
...
     box-shadow: ${parseDropShadow("drop-shadow-centered-lemon")};
   }
-  .yellow-shadow:has(+ .internal-button:hover) {
-    box-shadow: ${parseDropShadow("drop-shadow-lemon")};
-  }
+  ${(props) => css`
+    .yellow-shadow:has(+ .internal-button:hover) {
+      box-shadow: ${parseDropShadow(props.$hoverShadow)};
+    }
+  `};
```
Comment (carlmw) [inline src/components/ui/InternalShadowRectButton/InternalShadowRectButton.tsx:99]: 🏄🏼 minor but the value is the only bit that needs interpolating here.

```suggestion
  .yellow-shadow:has(+ .internal-button:hover) {
    box-shadow: ${(props) => parseDropShadow(props.$hoverShadow)};
  }
```

**Review by kimon-satan — CHANGES_REQUESTED** [review]
Not sure how you are still setting a background colors even though they don't render. It should be that they are set to null or none ?

File: src/components/ui/OakPrimaryInvertedButton/OakPrimaryInvertedButton.tsx (hunk 2 above)
Comment (weronika-szalas) [reply src/components/ui/OakPrimaryInvertedButton/OakPrimaryInvertedButton.tsx:32]: it is set to the same background colour in default state and on hover so it is not suppose to change

Comment (carlmw) [reply src/components/ui/OakPrimaryInvertedButton/OakPrimaryInvertedButton.tsx:32]: `bg-btn-secondary` is white isn't it?

Screenshot 2024-01-31 at 11 38 46

might be worth setting a default background on the story to make that clearer?

```typescript
  parameters: {
    backgrounds: {
      default: "light",
    },
  }
```

File: src/components/ui/InternalShadowRectButton/InternalShadowRectButton.tsx [hunk 5]
```diff
@@ -80,16 +81,23 @@ const StyledInternalButton = styled(InternalButton)<
...
   .grey-shadow:has(+ * + .internal-button:focus-visible) {
     box-shadow: ${parseDropShadow("drop-shadow-centered-grey")};
   }
   .yellow-shadow:has(+ .internal-button:focus-visible) {
     box-shadow: ${parseDropShadow("drop-shadow-centered-lemon")};
   }
-  .yellow-shadow:has(+ .internal-button:hover) {
-    box-shadow: ${parseDropShadow("drop-shadow-lemon")};
-  }
+  ${(props) => css`
```
Comment (carlmw) [inline src/components/ui/InternalShadowRectButton/InternalShadowRectButton.tsx:95]: Minor but this is messing with the focus styles, probably a matter of specificity

https://github.com/oaknational/oak-components/assets/122096/16793583-3318-45ec-8d85-dd6fa5bf8347

File: src/components/ui/OakPrimaryInvertedButton/OakPrimaryInvertedButton.tsx (hunk 3 above)
Comment (weronika-szalas) [reply src/components/ui/OakPrimaryInvertedButton/OakPrimaryInvertedButton.tsx:35]: the background is white, I can wrap in in a coloured box in Storybook to make it clear

**Review by carlmw — APPROVED** [review]
LGTM just some minor bits.

I use 🏄🏼 to point to comments that are just my opinion and you can choose to ignore them

File: src/components/ui/OakPrimaryInvertedButton/OakPrimaryInvertedButton.stories.tsx (hunk 1 above)
Comment (weronika-szalas) [reply src/components/ui/OakPrimaryInvertedButton/OakPrimaryInvertedButton.stories.tsx:23]: good shout, updated

File: src/components/ui/InternalShadowRectButton/InternalShadowRectButton.tsx (hunk 4 above)
Comment (weronika-szalas) [reply src/components/ui/InternalShadowRectButton/InternalShadowRectButton.tsx:99]: updated

File: src/components/ui/InternalShadowRectButton/InternalShadowRectButton.tsx (hunk 5 above)
Comment (weronika-szalas) [reply src/components/ui/InternalShadowRectButton/InternalShadowRectButton.tsx:95]: updated

**Review by kimon-satan — APPROVED** [review]

**Review by rightsaidjames — APPROVED** [review]
Bee-yoo-tee-full

**Pull request merged**